File: App-RentHouse/service/poi-service.js
// services/poi-service.js - TỐI ƯU HÓA
const axios = require('axios');

/**
 * 🗺️ POI Categories - Các danh mục địa điểm quan trọng
 */
const POI_CATEGORIES = {
  EDUCATION: {
    name: 'Giáo dục',
    icon: '🎓',
    tags: ['school', 'university', 'college', 'kindergarten'],
    query: 'node["amenity"~"school|university|college|kindergarten"]'
  },
  HOSPITAL: {
    name: 'Y tế',
    icon: '🏥',
    tags: ['hospital', 'clinic', 'doctors', 'pharmacy'],
    query: 'node["amenity"~"hospital|clinic|doctors|pharmacy"]'
  },
  TRANSPORT: {
    name: 'Giao thông',
    icon: '🚌',
    tags: ['bus_station', 'railway_station', 'subway_entrance'],
    query: 'node["amenity"~"bus_station|railway_station|subway_entrance"]'
  },
  SHOPPING: {
    name: 'Mua sắm',
    icon: '🛒',
    tags: ['supermarket', 'mall', 'marketplace'],
    query: 'node["shop"~"supermarket|mall|marketplace"]'
  },
  RESTAURANT: {
    name: 'Ăn uống',
    icon: '🍽️',
    tags: ['restaurant', 'cafe', 'fast_food'],
    query: 'node["amenity"~"restaurant|cafe|fast_food"]'
  },
  PARK: {
    name: 'Công viên',
    icon: '🌳',
    tags: ['park', 'garden'],
    query: 'node["leisure"~"park|garden"]'
  },
  BANK: {
    name: 'Ngân hàng',
    icon: '🏦',
    tags: ['bank', 'atm'],
    query: 'node["amenity"~"bank|atm"]'
  },
  GYM: {
    name: 'Thể thao',
    icon: '💪',
    tags: ['fitness_centre', 'sports_centre'],
    query: 'node["leisure"~"fitness_centre|sports_centre"]'
  }
};

/**
 * 🔍 Overpass API Service - Lấy POI từ OpenStreetMap
 * ✅ TỐI ƯU HÓA: Xử lý lỗi 504, giới hạn POI, retry logic
 */
class POIService {
  constructor() {
    this.overpassUrl = 'https://overpass-api.de/api/interpreter';
    this.cache = new Map();
    this.cacheDuration = 1000 * 60 * 30; // 30 minutes
    this.maxPOIsPerCategory = 50; // ✅ GIỚI HẠN 50 POI/category
    this.requestTimeout = 45000; // ✅ TIMEOUT 45s
    this.maxRetries = 2; // ✅ RETRY 2 lần nếu lỗi
  }

  /**
   * 🌍 Lấy POI theo category và vị trí
   * ✅ CẬP NHẬT: Retry logic + Timeout handling + POI limiting
   */
  async getPOIsByCategory(latitude, longitude, category, radiusKm = 5) {
    const cacheKey = `${category}_${latitude}_${longitude}_${radiusKm}`;
    
    // Check cache
    if (this.cache.has(cacheKey)) {
      const cached = this.cache.get(cacheKey);
      if (Date.now() - cached.timestamp < this.cacheDuration) {
        console.log(`✅ [POI-CACHE] Using cached data for ${category} (${cached.data.length} POIs)`);
        return cached.data;
      }
    }

    const categoryConfig = POI_CATEGORIES[category];
    if (!categoryConfig) {
      throw new Error(`Invalid category: ${category}`);
    }

    let lastError;
    
    // ✅ RETRY LOGIC: Thử lại tối đa maxRetries lần
    for (let attempt = 1; attempt <= this.maxRetries + 1; attempt++) {
      try {
        const radiusMeters = radiusKm * 1000;

        // ✅ CẬP NHẬT: Thêm [maxsize:...] để giới hạn response size
        const query = `
          [out:json][timeout:30][maxsize:536870912];
          (
            ${categoryConfig.query}(around:${radiusMeters},${latitude},${longitude});
          );
          out center;
        `;

        console.log(`🔍 [POI-SERVICE] Fetching ${category} (Attempt ${attempt}/${this.maxRetries + 1}) near (${latitude}, ${longitude}), radius: ${radiusKm}km`);

        const response = await axios.post(
          this.overpassUrl,
          `data=${encodeURIComponent(query)}`,
          {
            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
            timeout: this.requestTimeout
          }
        );

        const pois = this.parsePOIResponse(response.data, category, latitude, longitude);

        // ✅ GIỚI HẠN POIs: Lấy 50 gần nhất
        const limitedPOIs = pois
          .sort((a, b) => a.distanceFromUser - b.distanceFromUser)
          .slice(0, this.maxPOIsPerCategory);

        // Cache result
        this.cache.set(cacheKey, {
          data: limitedPOIs,
          timestamp: Date.now()
        });

        console.log(`✅ [POI-SERVICE] Found ${pois.length} ${category} POIs, limited to ${limitedPOIs.length}`);
        return limitedPOIs;

      } catch (error) {
        lastError = error;
        
        if (error.response?.status === 504 || error.code === 'ECONNABORTED') {
          console.warn(`⚠️ [POI-SERVICE] Attempt ${attempt} failed (${error.response?.status || error.code}), retrying...`);
          
          // ✅ BACKOFF: Chờ 2s trước khi retry
          if (attempt <= this.maxRetries) {
            await new Promise(resolve => setTimeout(resolve, 2000));
            continue;
          }
        } else {
          // Lỗi khác, không retry
          break;
        }
      }
    }

    // ✅ FALLBACK: Nếu tất cả retry thất bại, trả về array rỗng thay vì throw error
    console.error(`❌ [POI-SERVICE] Error fetching ${category} after ${this.maxRetries + 1} attempts:`, lastError.message);
    return []; // Trả về array rỗng thay vì crash
  }

  /**
   * 📊 Parse Overpass API response
   */
  parsePOIResponse(data, category, userLat, userLon) {
    if (!data || !data.elements || data.elements.length === 0) {
      return [];
    }

    const categoryConfig = POI_CATEGORIES[category];

    return data.elements
      .filter(element => element.lat && element.lon) // ✅ Filter invalid coords
      .map(element => {
        const poiLat = element.lat;
        const poiLon = element.lon;
        
        // 🔥 TÍNH KHOẢNG CÁCH TỪ USER ĐẾN POI
        const distanceFromUser = this.calculateDistance(
          userLat,
          userLon,
          poiLat,
          poiLon
        );

        return {
          id: `poi_${category}_${element.id}`,
          osmId: element.id,
          category: category,
          categoryName: categoryConfig.name,
          categoryIcon: categoryConfig.icon,
          name: element.tags?.name || `${categoryConfig.name} (không tên)`,
          latitude: poiLat,
          longitude: poiLon,
          tags: element.tags || {},
          distanceFromUser: distanceFromUser,
          address: this.buildAddress(element.tags)
        };
      });
  }

  /**
   * 🏠 Build address from OSM tags
   */
  buildAddress(tags) {
    const parts = [];
    if (tags['addr:housenumber']) parts.push(tags['addr:housenumber']);
    if (tags['addr:street']) parts.push(tags['addr:street']);
    if (tags['addr:district']) parts.push(tags['addr:district']);
    if (tags['addr:city']) parts.push(tags['addr:city']);
    return parts.join(', ') || 'Không có địa chỉ';
  }

  /**
   * 🎯 Lấy tất cả POI categories gần vị trí
   */
  async getAllPOIsNearby(latitude, longitude, radiusKm = 5) {
    try {
      const results = {};
      const categories = Object.keys(POI_CATEGORIES);

      console.log(`🌍 [POI-SERVICE] Fetching all categories near (${latitude}, ${longitude})`);

      const promises = categories.map(async (category) => {
        const pois = await this.getPOIsByCategory(latitude, longitude, category, radiusKm);
        results[category] = pois;
      });

      await Promise.all(promises);

      const totalPOIs = Object.values(results).reduce((sum, pois) => sum + pois.length, 0);
      console.log(`✅ [POI-SERVICE] Total POIs found: ${totalPOIs}`);

      return results;

    } catch (error) {
      console.error('❌ [POI-SERVICE] Error fetching all POIs:', error.message);
      return {};
    }
  }

  /**
   * 🧮 Haversine formula - Calculate distance between 2 coordinates
   */
  calculateDistance(lat1, lon1, lat2, lon2) {
    const R = 6371; // Earth radius in km
    const dLat = this.toRadians(lat2 - lat1);
    const dLon = this.toRadians(lon2 - lon1);

    const a =
      Math.sin(dLat / 2) * Math.sin(dLat / 2) +
      Math.cos(this.toRadians(lat1)) *
      Math.cos(this.toRadians(lat2)) *
      Math.sin(dLon / 2) *
      Math.sin(dLon / 2);

    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    return R * c;
  }

  toRadians(degrees) {
    return degrees * (Math.PI / 180);
  }

  /**
   * 🔥 Filter rentals based on POI distance
   */
  filterRentalsByPOIDistance(poi, rentals, radiusKm = 3) {
    return rentals.map(rental => {
      const rentalLat = rental.location?.coordinates?.coordinates?.[1] || 0;
      const rentalLon = rental.location?.coordinates?.coordinates?.[0] || 0;

      if (rentalLat === 0 || rentalLon === 0) {
        return { ...rental, distanceFromPOI: Infinity, withinRadius: false };
      }

      const distance = this.calculateDistance(
        poi.latitude,
        poi.longitude,
        rentalLat,
        rentalLon
      );

      return {
        ...rental,
        distanceFromPOI: distance,
        withinRadius: distance <= radiusKm,
        nearbyPOI: {
          name: poi.name,
          category: poi.categoryName,
          icon: poi.categoryIcon,
          distance: this.formatDistance(distance)
        }
      };
    }).filter(r => r.withinRadius);
  }

  /**
   * 🔥 Filter rentals by multiple POI categories
   */
  filterRentalsByMultiplePOIs(selectedCategories, poiData, rentals, radiusKm = 3) {
    const filteredRentals = [];
    const processedRentalIds = new Set();
    
    // 🔥 BƯỚC 1: THU THẬP TẤT CẢ POI TỪ CÁC CATEGORIES ĐÃ CHỌN
    const allPOIs = [];
    
    for (const categoryId of selectedCategories) {
      const pois = poiData[categoryId] || [];
      allPOIs.push(...pois);
    }
    
    console.log(`📊 [FILTER-POI] Total POIs from ${selectedCategories.length} categories: ${allPOIs.length}`);
    
    // ✅ CẬP NHẬT: Không còn lấy 100 POI, mà dùng tất cả POI đã giới hạn
    // (mỗi category tối đa 50 POI, vậy tối đa 100 POI nếu 2 categories)
    allPOIs.sort((a, b) => a.distanceFromUser - b.distanceFromUser);
    
    console.log(`✂️ [FILTER-POI] Using ${allPOIs.length} POIs for filtering`);
    
    // 🔥 BƯỚC 2: XỬ LÝ VỚI TẤT CẢ POI
    for (const poi of allPOIs) {
      const rentalsNearPOI = this.filterRentalsByPOIDistance(poi, rentals, radiusKm);
  
      for (const rental of rentalsNearPOI) {
        if (!processedRentalIds.has(rental._id.toString())) {
          rental.nearestPOIs = [{
            name: poi.name,
            category: poi.categoryName,
            icon: poi.categoryIcon,
            distance: this.formatDistance(rental.distanceFromPOI)
          }];
          
          filteredRentals.push(rental);
          processedRentalIds.add(rental._id.toString());
        } else {
          const existingRental = filteredRentals.find(r => r._id.toString() === rental._id.toString());
          if (existingRental && !existingRental.nearestPOIs.some(p => p.name === poi.name)) {
            existingRental.nearestPOIs.push({
              name: poi.name,
              category: poi.categoryName,
              icon: poi.categoryIcon,
              distance: this.formatDistance(rental.distanceFromPOI)
            });
          }
        }
      }
    }
  
    // ✅ Sort rentals theo khoảng cách gần nhất
    return filteredRentals.sort((a, b) => {
      const minDistA = Math.min(...(a.nearestPOIs?.map(p => parseFloat(p.distance)) || [Infinity]));
      const minDistB = Math.min(...(b.nearestPOIs?.map(p => parseFloat(p.distance)) || [Infinity]));
      return minDistA - minDistB;
    });
  }
  
  /**
   * 🔥 FORMAT DISTANCE
   */
  formatDistance(distanceKm) {
    if (distanceKm < 1) {
      return (distanceKm * 1000).toFixed(0); // Trả về mét (string)
    }
    return distanceKm.toFixed(2); // Trả về km (string)
  }

  /**
   * 🗑️ Clear cache
   */
  clearCache() {
    this.cache.clear();
    console.log('✅ POI cache cleared');
  }
}

module.exports = {
  POIService,
  POI_CATEGORIES
};

File: App_Flutter_rentalhouse/lib/services/poi_service.dart
import 'dart:convert';
import 'dart:math'; // ✅ FIX: Import dart:math
import 'package:http/http.dart' as http;
import 'package:flutter/foundation.dart';
import '../config/api_routes.dart';
import '../models/poi.dart';
import '../models/rental.dart';
import '../services/auth_service.dart';

class POIService {
  static final POIService _instance = POIService._internal();
  factory POIService() => _instance;
  POIService._internal();

  final http.Client _client = http.Client();
  static const Duration _timeout = Duration(seconds: 30);

  /// 🏷️ Lấy danh sách categories
  Future<List<POICategory>> getCategories() async {
    try {
      debugPrint('🔍 [POI-SERVICE] Fetching categories');

      final response = await _client
          .get(Uri.parse(ApiRoutes.poiCategories))
          .timeout(_timeout);

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final List<dynamic> categoriesData = data['categories'] ?? [];

        debugPrint('✅ [POI-SERVICE] Found ${categoriesData.length} categories');

        return categoriesData
            .map((json) => POICategory.fromJson(json))
            .toList();
      } else {
        throw Exception('Failed to get categories: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('❌ [POI-SERVICE] Error getting POI categories: $e');
      rethrow;
    }
  }
  Future<Map<String, dynamic>> getAIPersonalizedWithPOI({
    required double latitude,
    required double longitude,
    required List<String> selectedCategories,
    double radius = 10.0,
    double poiRadius = 3.0,
    double? minPrice,
    double? maxPrice,
    int? limit,
  }) async {
    try {
      final token = await AuthService().getIdToken();

      if (token == null) {
        throw Exception('Vui lòng đăng nhập để xem gợi ý AI + POI');
      }

      final body = {
        'latitude': latitude,
        'longitude': longitude,
        'selectedCategories': selectedCategories,
        'radius': radius,
        'poiRadius': poiRadius,
        if (limit != null) 'limit': limit,
        if (minPrice != null) 'minPrice': minPrice,
        if (maxPrice != null) 'maxPrice': maxPrice,
      };

      debugPrint('🤖🏢 [AI+POI-SERVICE] Request:');
      debugPrint('   Categories: ${selectedCategories.join(", ")}');
      debugPrint('   Radius: ${radius}km, POI Radius: ${poiRadius}km');
      debugPrint('   Coordinates: ($latitude, $longitude)');

      final response = await _client.post(
        Uri.parse(ApiRoutes.aiPOIRecommendations),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: jsonEncode(body),
      ).timeout(_timeout);

      debugPrint('🤖🏢 [AI+POI-SERVICE] Response: ${response.statusCode}');

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);

        if (data['success'] != true) {
          throw Exception(data['message'] ?? 'API returned error');
        }

        // ✅ Parse rentals
        final List<dynamic> rentalsData = data['rentals'] ?? [];
        final rentals = rentalsData
            .map((json) => Rental.fromJson(json))
            .toList();

        debugPrint('🤖🏢 [AI+POI-SERVICE] Success:');
        debugPrint('   Rentals: ${rentals.length}');
        debugPrint('   POIs found: ${data['poiStats']?['totalPOIsFound'] ?? 0}');
        debugPrint('   Method: ${data['method']}');

        return {
          'rentals': rentals,
          'total': data['total'] ?? rentals.length,
          'poisTotal': data['poiStats']?['totalPOIsFound'] ?? 0,
          'selectedCategories': selectedCategories,
          'message': data['message'] ?? 'Gợi ý AI + POI',
          'method': data['method'] ?? 'ai_poi_combined',
          'success': true,
        };
      } else {
        throw Exception('Failed: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('❌ [AI+POI-SERVICE] Error: $e');
      return {
        'rentals': <Rental>[],
        'total': 0,
        'poisTotal': 0,
        'selectedCategories': selectedCategories,
        'message': 'Lỗi: $e',
        'success': false,
      };
    }
  }
  /// 📍 Lấy POI theo category và vị trí
  Future<List<POI>> getPOIsNearby({
    required double latitude,
    required double longitude,
    String? category,
    double radius = 5.0,
  }) async {
    try {
      final url = ApiRoutes.poiNearby(
        latitude: latitude,
        longitude: longitude,
        category: category,
        radius: radius,
      );

      debugPrint('🔍 [POI-SERVICE] Fetching POIs: category=$category, radius=${radius}km');

      final response = await _client
          .get(Uri.parse(url))
          .timeout(_timeout);

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final List<dynamic> poisData = data['pois'] ?? [];

        final pois = poisData.map((json) => POI.fromJson(json)).toList();

        debugPrint('✅ [POI-SERVICE] Found ${pois.length} POIs');
        return pois;
      } else {
        throw Exception('Failed to get POIs: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('❌ [POI-SERVICE] Error getting POIs: $e');
      return [];
    }
  }

  /// 🔥 Lọc rentals theo POI và khoảng cách
  /// ✅ FIX: Sử dụng endpoint đúng từ ApiRoutes
  Future<POIFilterResult> filterRentalsByPOI({
    required double latitude,
    required double longitude,
    required List<String> selectedCategories,
    double radius = 3.0,
    double? minPrice,
    double? maxPrice,
    int? limit,
  }) async {
    try {
      final token = await AuthService().getIdToken();

      if (token == null) {
        throw Exception('Vui lòng đăng nhập');
      }

      final body = {
        'latitude': latitude,
        'longitude': longitude,
        'selectedCategories': selectedCategories,
        'radius': radius,
        if (limit != null) 'limit': limit,
        if (minPrice != null) 'minPrice': minPrice,
        if (maxPrice != null) 'maxPrice': maxPrice,
      };

      debugPrint('🔥 [POI-FILTER] Request: categories=${selectedCategories.join(", ")}, radius=${radius}km');

      final response = await _client.post(
        Uri.parse(ApiRoutes.filterRentalsByPOI),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: jsonEncode(body),
      ).timeout(_timeout);

      debugPrint('🔥 [POI-FILTER] Response status: ${response.statusCode}');

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);

        if (data['success'] != true) {
          throw Exception(data['message'] ?? 'API returned error');
        }

        // ✅ FIX: Parse rentals (remain as dynamic list)
        final List<dynamic> rentalsData = data['rentals'] ?? [];
        final rentals = rentalsData
            .map((json) => Rental.fromJson(json))
            .toList();

        // ✅ FIX: Convert pois từ Map<String, dynamic> sang POIOnMap
        final List<dynamic> poisData = data['pois'] ?? [];
        final pois = poisData
            .whereType<Map<String, dynamic>>()
            .map((poiJson) => POIOnMap.fromJson(poiJson))
            .toList();

        final result = POIFilterResult(
          rentals: rentals,
          pois: pois, // ✅ FIX: Giờ là List<POIOnMap>
          total: data['total'] as int? ?? rentals.length,
          poisTotal: data['poisTotal'] as int? ?? 0,
          selectedCategories: selectedCategories,
          radius: radius,
          message: data['message'] ?? 'Tìm thấy bài gần tiện ích',
          success: true,
        );

        debugPrint('✅ [POI-FILTER] Found ${result.total} rentals, ${result.poisTotal} POIs');
        return result;
      } else {
        throw Exception('Failed: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('❌ [POI-FILTER] Error: $e');
      return POIFilterResult(
        rentals: [],
        pois: [], // ✅ Empty list<POIOnMap>
        total: 0,
        poisTotal: 0,
        selectedCategories: selectedCategories,
        radius: radius,
        message: 'Lỗi: $e',
        success: false,
      );
    }
  }
  /// 🏠 Lấy rentals gần POI
  Future<Map<String, dynamic>> getRentalsNearPOI({
    required POI poi,
    double radius = 5.0,
    double? minPrice,
    double? maxPrice,
    int? limit,
  }) async {
    try {
      final token = await AuthService().getIdToken();

      final body = {
        'poi': poi.toJson(),
        'radius': radius,
        if (limit != null) 'limit': limit,
        if (minPrice != null) 'minPrice': minPrice,
        if (maxPrice != null) 'maxPrice': maxPrice,
      };

      debugPrint('🏠 [POI-SERVICE] Fetching rentals near ${poi.name}');

      final response = await _client.post(
        Uri.parse(ApiRoutes.rentalsNearPOI),
        headers: {
          'Content-Type': 'application/json',
          if (token != null) 'Authorization': 'Bearer $token',
        },
        body: jsonEncode(body),
      ).timeout(_timeout);

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final List<dynamic> rentalsData = data['rentals'] ?? [];

        final rentals = rentalsData
            .map((json) => Rental.fromJson(json))
            .toList();

        debugPrint('✅ [POI-SERVICE] Found ${rentals.length} rentals');

        return {
          'rentals': rentals,
          'total': data['total'] ?? 0,
          'poi': poi,
          'success': true,
        };
      } else {
        throw Exception('Failed: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('❌ [POI-SERVICE] Error: $e');
      return {
        'rentals': <Rental>[],
        'total': 0,
        'poi': poi,
        'success': false,
        'error': e.toString(),
      };
    }
  }

  /// 🗺️ Convert POI to POIOnMap
  POIOnMap toPOIOnMap(POI poi, {required bool hasNearbyRentals}) {
    return POIOnMap.fromPOI(poi, hasNearbyRentals: hasNearbyRentals);
  }

  /// 🎯 Filter POIs by category
  List<POI> filterPOIsByCategory(List<POI> allPOIs, String category) {
    return allPOIs.where((poi) => poi.category == category).toList();
  }

  /// 📏 Calculate distance (Haversine formula)
  /// ✅ FIX: Use dart:math functions
  double calculateDistance(double lat1, double lon1, double lat2, double lon2) {
    const R = 6371; // Earth radius in km
    final dLat = _toRadians(lat2 - lat1);
    final dLon = _toRadians(lon2 - lon1);

    final a = sin(dLat / 2) * sin(dLat / 2) +
        cos(_toRadians(lat1)) *
            cos(_toRadians(lat2)) *
            sin(dLon / 2) *
            sin(dLon / 2);

    final c = 2 * atan2(sqrt(a), sqrt(1 - a));
    return R * c;
  }

  double _toRadians(double degrees) {
    return degrees * (pi / 180); // ✅ Use pi from dart:math
  }

  // ✅ FIX: Phần getAIPOIRecommendations trong poi_service.dart

  /// 🤖 Get AI + POI combined recommendations
  Future<Map<String, dynamic>> getAIPOIRecommendations({
    required double latitude,
    required double longitude,
    required List<String> selectedCategories,
    double radius = 10.0,
    double? minPrice,
    double? maxPrice,
    int? limit,
  }) async {
    try {
      final token = await AuthService().getIdToken();

      if (token == null) {
        throw Exception('Vui lòng đăng nhập để xem gợi ý AI');
      }

      final body = {
        'latitude': latitude,
        'longitude': longitude,
        'selectedCategories': selectedCategories,
        'radius': radius,
        if (limit != null) 'limit': limit,
        if (minPrice != null) 'minPrice': minPrice,
        if (maxPrice != null) 'maxPrice': maxPrice,
      };

      debugPrint('🤖 [AI+POI] Request: categories=${selectedCategories.join(", ")}, radius=${radius}km');

      final response = await _client.post(
        Uri.parse(ApiRoutes.aiPOIRecommendations),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: jsonEncode(body),
      ).timeout(_timeout);

      debugPrint('🤖 [AI+POI] Response status: ${response.statusCode}');

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);

        if (data['success'] != true) {
          throw Exception(data['message'] ?? 'API returned error');
        }

        // ✅ FIX: Parse rentals correctly
        final List<dynamic> rentalsData = data['rentals'] ?? [];
        final rentals = rentalsData
            .map((json) => Rental.fromJson(json))
            .toList();

        debugPrint('✅ [AI+POI] Found ${rentals.length} rentals');

        return {
          'rentals': rentals, // ✅ List<Rental>
          'isAIRecommendation': data['isAIRecommendation'] ?? true,
          'message': data['message'] ?? 'Gợi ý AI + POI',
          'total': data['total'] ?? rentals.length,
          'poisTotal': data['poisFound'] ?? 0,
          'success': true,
        };
      } else {
        throw Exception('Failed: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('❌ [AI+POI] Error: $e');
      return {
        'rentals': <Rental>[],
        'isAIRecommendation': false,
        'message': 'Lỗi: $e',
        'total': 0,
        'poisTotal': 0,
        'success': false,
      };
    }
  }
  void dispose() {
    _client.close();
  }
}